### DISCUSSION TITLE
DynJS.axd LocalText sometime return 404

### DISCUSSION THREAD

**Discussion by minhhungit** [serenity-is/Serenity#6513] [Help]
hi, latest serenity, DynJS.axd LocalText some time return 404, does anyone get this before, any suggestion pls 

![dynjs local text 404 2](https://user-images.githubusercontent.com/2279508/186497070-f1858592-e1c4-421a-8422-3277fd581a41.gif)

**Comment by minhhungit** [comment]
it looks just happen when deploy website on IIS, local looks work without the problem

**Comment by minhhungit** [comment]
it looks it happen when we deploy the website with multi IIS worker (iis web garden)

**Reply by minhhungit** [reply]
this is my IIS config

![image](https://user-images.githubusercontent.com/2279508/186528153-4010b184-0259-4da0-ab1a-9f3ef36d6479.png)

**Reply by VictorTomaili** [reply]
> it looks it happen when we deploy the website with multi IIS worker (iis web garden)

In this case, this means you have a load balancer in front of your application. When you send the request, its retrieve data from different instances. One of your instance has misconfiguration. And other one is working well.

**Reply by minhhungit** [reply]
no @VictorTomaili this is on Test server and the server have no load balancer

> When you send the request, its retrieve data from different instances.

yes, maybe. I think I understood the problem, I guess it due to when we refresh DynJS.axd url, the url is processed on a new IIS worker **before localtext data is initialized on the new worker.**

Maybe we need to check and try to init data for DynJS LocalText if it has not initialized yet - before we return 404 to client

Btw, on production server, I also have a load balancer front of 2 web server, and IIS on each server is also configured with multi IIS workers, and I also see the same problem.

**Reply by VictorTomaili** [reply]
The version will be different for the instances

**Reply by minhhungit** [reply]
@VictorTomaili is there a way we can config DynJS to use distributed caching instead memory cache

**Reply by VictorTomaili** [reply]
The script version is not cache related. And when I check the version, you can write anything to the version, it's working even you modify the version number manually. Probably, one of the instance has wrong configuration. 

The script must use the local cache, because if you deploy the new version, it must use new version of the data per instance. Because of that, the dynjs uses the in memory cache. With the distributed cache, after deploy the new instance, your new instance can try to use the old script, or the old instance can try to use new script. It's not recommended.

**Reply by minhhungit** [reply]
> The script version is not cache related. And when I check the version, you can write anything to the version

Agree with you

> Probably, one of the instance has wrong configuration.

They are same, because only one deploy package, one folder, IIS will auto create iis worker for us 

> The script must use the local cache

Yes local cache for each IIS worker (instance). But we need to make sure that when we run new instance, it must re-cache new data, right?

I have a suggestion, we can declare a static guid value everytime we start web app, and because static variable will not be shared between workers in same app pool so the value will be difference for every instance, and we use the guid value as a part of cache key, for example:
```csharp
public static class InstanceData{
     public static Guid SessionId = Guid.NewGuid();
}

var masterCacheKey = InstanceData.SessionId + cachekey
```

**Reply by VictorTomaili** [reply]
In this case, you lost the advantage of the distributed cache. Use in memory cache and use distributed cache with the instance key are nearly same thing. 
Check your deployment again, there is no possibility to randomized not found error. This issue is not related to the in memory or distributed cache. Use same folder for 2 IIS worker can cause this error, but I'm not ensuring. Good luck.

**Reply by minhhungit** [reply]
thank you @VictorTomaili I will check again

**Reply by minhhungit** [reply]
just note, some time for same url + version (when run app in IIS web garden with multiple workers), the url also return 404
for example this version: `v=C5TmIrRSo7v-pejpnVj76Q`, it was 200 many times before I got 404, after 404 I got 200 again

![image](https://user-images.githubusercontent.com/2279508/188134714-f04dc599-94a0-4767-b4b7-b25ba71d2aa9.png)

I tried to check but it's very hard to debug because hard to control IIS worker
almost I don't get the problem when run with only 1 worker 

I'm quite sure this is not problem of deploying, because all workers using same app folder, same IIS, same machine
Maybe I addressed wrong reason (caching I mean) but quite sure problem happened with multi workers

**Reply by minhhungit** [reply]
here is more info, hope it can give you guys some clues 

![image](https://user-images.githubusercontent.com/2279508/188138379-85140573-2695-469b-ada7-b78cd1310f8b.png)

![image](https://user-images.githubusercontent.com/2279508/188138576-c70c67b6-1f4f-4695-88f5-665c607941f0.png)
